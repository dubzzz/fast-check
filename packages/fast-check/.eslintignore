bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs