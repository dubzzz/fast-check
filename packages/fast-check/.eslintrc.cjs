module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
  ],
  rules: {
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/no-empty-function': 'warn',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-this-alias': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/no-use-before-define': 'warn',
  },
  overrides: [
    {
      files: ['*.spec.ts', 'test/unit/**/*.ts', 'test/e2e/**/*.ts'],
      rules: {
        '@typescript-eslint/no-empty-function': 'off',
        '@typescript-eslint/no-use-before-define': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
      },
    },
  ],
};
