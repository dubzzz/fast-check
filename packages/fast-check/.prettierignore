bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/