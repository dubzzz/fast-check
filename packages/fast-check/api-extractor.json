{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "compiler": {
    "tsconfigFilePath": "<projectFolder>/tsconfig.publish.types.json"
  },
  "tsdocMetadata": {
    "enabled": false
  },
  "mainEntryPointFilePath": "lib/types/fast-check-default.d.ts",
  "dtsRollup": {
    "enabled": false
  },
  "docModel": {
    "enabled": true,
    "apiJsonFilePath": "docs/fast-check.api.json"
  },
  "apiReport": {
    "enabled": false
  },
  "messages": {
    "compilerMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    },
    "extractorMessageReporting": {
      "default": {
        "logLevel": "error"
      },
      "ae-unresolved-link": {
        "logLevel": "warning"
      }
    },
    "tsdocMessageReporting": {
      "default": {
        "logLevel": "error"
      }
    }
  }
}
