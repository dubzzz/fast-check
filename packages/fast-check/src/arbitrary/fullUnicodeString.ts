import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { fullUnicode } from './fullUnicode';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

const safeObjectAssign = Object.assign;

/**
 * For strings of {@link fullUnicode}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.11
 * @public
 */
export function fullUnicodeString(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const charArbitrary = fullUnicode();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, codePointsToStringUnmapper);
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const enrichedConstraints: ArrayConstraintsInternal<string> = safeObjectAssign(safeObjectAssign({}, constraints), {
    experimentalCustomSlices,
  });
  return array(charArbitrary, enrichedConstraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
