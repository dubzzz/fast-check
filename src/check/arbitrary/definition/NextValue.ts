import { cloneMethod, hasCloneMethod, WithCloneMethod } from '../../symbols';

/**
 * A `NextValue<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 2.15.0
 * @public
 */
export class NextValue<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown = undefined, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = hasCloneMethod(value_);
    this.readOnce = false;
    Object.defineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return ((this.value_ as unknown) as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
