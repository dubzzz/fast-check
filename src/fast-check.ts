import * as fc from './fast-check-default';
export default fc;

export * from './fast-check-default';
