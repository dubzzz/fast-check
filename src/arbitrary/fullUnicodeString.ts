import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { fullUnicode } from './fullUnicode';
import {
  buildStringArbitrary,
  StringFullConstraintsDefinition,
  StringSharedConstraints,
} from './_internals/builders/StringArbitraryBuilder';
export { StringSharedConstraints } from './_internals/builders/StringArbitraryBuilder';

/**
 * For strings of {@link fullUnicode}
 * @remarks Since 0.0.11
 * @public
 */
function fullUnicodeString(): Arbitrary<string>;
/**
 * For strings of {@link fullUnicode}
 *
 * @param maxLength - Upper bound of the generated string length
 *
 * @deprecated
 * Superceded by `fc.fullUnicodeString({maxLength})` - see {@link https://github.com/dubzzz/fast-check/issues/992 | #992}.
 * Ease the migration with {@link https://github.com/dubzzz/fast-check/tree/main/codemods/unify-signatures | our codemod script}.
 *
 * @remarks Since 0.0.11
 * @public
 */
function fullUnicodeString(maxLength: number): Arbitrary<string>;
/**
 * For strings of {@link fullUnicode}
 *
 * @param minLength - Lower bound of the generated string length
 * @param maxLength - Upper bound of the generated string length
 *
 * @deprecated
 * Superceded by `fc.fullUnicodeString({minLength, maxLength})` - see {@link https://github.com/dubzzz/fast-check/issues/992 | #992}.
 * Ease the migration with {@link https://github.com/dubzzz/fast-check/tree/main/codemods/unify-signatures | our codemod script}.
 *
 * @remarks Since 0.0.11
 * @public
 */
function fullUnicodeString(minLength: number, maxLength: number): Arbitrary<string>;
/**
 * For strings of {@link fullUnicode}
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 2.4.0
 * @public
 */
function fullUnicodeString(constraints: StringSharedConstraints): Arbitrary<string>;
function fullUnicodeString(...args: StringFullConstraintsDefinition): Arbitrary<string> {
  return buildStringArbitrary(fullUnicode(), ...args);
}
export { fullUnicodeString };
