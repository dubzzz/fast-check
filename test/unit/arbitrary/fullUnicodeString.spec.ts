import * as fc from '../../../lib/fast-check';
import { fullUnicodeString } from '../../../src/arbitrary/fullUnicodeString';

import { NextValue } from '../../../src/check/arbitrary/definition/NextValue';
import {
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
} from './__test-helpers__/NextArbitraryAssertions';
import { buildNextShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

describe('fullUnicodeString (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(fc.nat({ max: 5 }), fc.nat({ max: 30 }), fc.boolean(), fc.boolean())
    .map(([min, gap, withMin, withMax]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
    }));

  const isCorrect = (value: string, extra: Extra) => {
    if (extra.minLength !== undefined) {
      expect([...value].length).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect([...value].length).toBeLessThanOrEqual(extra.maxLength);
    }
    for (const c of [...value]) {
      if (c.length === 1) {
        const isSurrogate = c.charCodeAt(0) >= 0xd800 && c.charCodeAt(0) <= 0xdfff;
        expect(isSurrogate).toBe(false);
      } else {
        const firstIsHighSurrogate = c.charCodeAt(0) >= 0xd800 && c.charCodeAt(0) <= 0xdbff;
        const secondIsLowSurrogate = c.charCodeAt(1) >= 0xdc00 && c.charCodeAt(1) <= 0xdfff;
        expect(firstIsHighSurrogate).toBe(true);
        expect(secondIsLowSurrogate).toBe(true);
      }
    }
  };

  const fullUnicodeStringBuilder = (extra: Extra) => fullUnicodeString(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(fullUnicodeStringBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(fullUnicodeStringBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(fullUnicodeStringBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(fullUnicodeStringBuilder, { extraParameters });
  });

  it.each`
    source
    ${''}
    ${'azerty'}
    ${'ah! ah!'}
    ${'a\uD83D\uDC31b'}
    ${'\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}\u{1f431}' /* by default maxLength = maxLengthFromMinLength(0) = 10 */}
    ${'\u{1f468}\u{1f3fe}\u{200d}\u{1f469}\u{1f3fc}'}
    ${'\u{1f468}\u{1f3fe}\u{200d}' /* accept incomplete graphemes */}
  `('should be able to generate $source with fc.fullUnicodeString()', ({ source }) => {
    // Arrange / Act
    const arb = fullUnicodeString();
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(true);
  });

  it.each`
    source                             | constraints
    ${'a\uD83Db' /* invalid string */} | ${{}}
    ${'ab' /* not large enough */}     | ${{ minLength: 3 }}
    ${'abcd' /* too large */}          | ${{ maxLength: 3 }}
  `('should not be able to generate $source with fc.fullUnicodeString($constraints)', ({ source, constraints }) => {
    // Arrange / Act
    const arb = fullUnicodeString(constraints);
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(false);
  });

  it.each`
    rawValue
    ${'Hey \u{1f431}!'}
    ${'\u{1f431}'.repeat(50) /* longer than default maxGeneratedLength but ok for shrink */}
  `('should be able to shrink $rawValue with fc.fullUnicodeString()', ({ rawValue }) => {
    // Arrange
    const arb = fullUnicodeString();
    const value = new NextValue(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildNextShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
