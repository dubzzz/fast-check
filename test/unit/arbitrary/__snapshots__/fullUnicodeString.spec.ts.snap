// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fullUnicodeString (integration) should be able to shrink Hey 🐱! 1`] = `
"\\"Hey 🐱!\\"
├> \\"\\"
├> \\" 🐱!\\"
|  ├> \\"🐱!\\"
|  |  ├> \\"!\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\" \\"
|  |  |     └> \\"\\"
|  |  ├> \\" !\\"
|  |  |  ├> \\"!\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\" \\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\" \\"
|  |  |  |  └> \\"\\"
|  |  |  └> \\"  \\"
|  |  |     ├> \\" \\"
|  |  |     |  └> \\"\\"
|  |  |     └> \\" \\"
|  |  |        └> \\"\\"
|  |  ├> \\"︙!\\"
|  |  |  ├> \\"!\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\" \\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"笍!\\"
|  |  |  |  ├> \\"!\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\" \\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  ├> \\"㶇!\\"
|  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  ├> \\"Ễ!\\"
|  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  ├> \\"ར!\\"
|  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  ├> \\"ޱ!\\"
|  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  ├> \\"ϙ!\\"
|  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  ├> \\"ǭ!\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"÷!\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\u001d!\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"^!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"?!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"0!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"(!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"$!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\\\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
